src/NuGet.Clients/NuGet.VisualStudio.Common/Services/NuGetFeatureFlagService.cs

```diff
@@ -0,0 +1,66 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.Composition;
+using System.Threading.Tasks;
+using Microsoft.Internal.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Shell;
+using NuGet.Common;
+
+namespace NuGet.VisualStudio
+{
+    [Export(typeof(INuGetFeatureFlagService))]
+    [PartCreationPolicy(CreationPolicy.Shared)]
+    public class NuGetFeatureFlagService : INuGetFeatureFlagService
+    {
+        private readonly IEnvironmentVariableReader _environmentVariableReader;
+        private readonly IAsyncServiceProvider _asyncServiceProvider;
+        private readonly Microsoft.VisualStudio.Threading.AsyncLazy<IVsFeatureFlags> _ivsFeatureFlags;
+        private readonly Dictionary<string, bool> _featureFlagCache;
+
+        [ImportingConstructor]
+        public NuGetFeatureFlagService()
+            : this(EnvironmentVariableWrapper.Instance, AsyncServiceProvider.GlobalProvider)
+        {
+            // ensure uniqueness.
+        }
+
+        internal NuGetFeatureFlagService(IEnvironmentVariableReader environmentVariableReader, IAsyncServiceProvider asyncServiceProvider)
+        {
+            _environmentVariableReader = environmentVariableReader ?? throw new ArgumentNullException(nameof(environmentVariableReader));
+            _asyncServiceProvider = asyncServiceProvider ?? throw new ArgumentNullException(nameof(asyncServiceProvider));
+            _ivsFeatureFlags = new(() => _asyncServiceProvider.GetServiceAsync<SVsFeatureFlags, IVsFeatureFlags>(), NuGetUIThreadHelper.JoinableTaskFactory);
+            _featureFlagCache = new();
+        }
+
+        public async Task<bool> IsFeatureEnabledAsync(NuGetFeatureFlagConstants featureFlag)
+        {
+            GetEnvironmentVariablesForFeature(featureFlag, out bool isFeatureForcedEnabled, out bool isFeatureForcedDisabled);
+            // Perform the check from the feature flag service only once.
+            // There are events sent for targetted notification changes, but we don't listen to those at this point.
+            if (!_featureFlagCache.TryGetValue(featureFlag.Name, out bool featureEnabled))
+            {
+                var featureFlagService = await _ivsFeatureFlags.GetValueAsync();
+                await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+                featureEnabled = featureFlagService.IsFeatureEnabled(featureFlag.Name, defaultValue: featureFlag.DefaultState);
+                _featureFlagCache.Add(featureFlag.Name, featureEnabled);
+            }
+            return !isFeatureForcedDisabled && (isFeatureForcedEnabled || featureEnabled);
+        }
+
+        private void GetEnvironmentVariablesForFeature(NuGetFeatureFlagConstants featureFlag, out bool isFeatureForcedEnabled, out bool isFeatureForcedDisabled)
```

Comment:
Can we share this code with the ExperimentationService? Whether it be through a separate object used by both or through inheritance of a base class.

Reply:
I think the base class approach is inappropriate for this mostly because they don't really share any content right now.
I'm not even sure what I would name it. 

We could have the same class implement both interfaces? 
Alternatively this could just be a static method. 

let me know what you think. 

Personally I'm not too worried about the duplicated code as it hasn't really met the 3 scenarios for reuse (there's 2 now).

Reply:
This is merged now so that's fine but for the sake of continuing the discussion, the way I would approach this is to have an object that specializes in providing an environment variable override. That object could be used by both services and the implementation of their "IsFeatureEnabled" or "IsExperimentEnabled" would delegate to the environment override object and depending on the result, decide whether to return that value or evaluate the feature flag / experiment service.

Said object that handles environment variable overrides would then be reusable for anything. It would be generalized enough and specialized in doing its one job well, which is to check if there is an environment variable that is set, and if so, if it set to true, or if it set to false. This object is unit testable and can have the environment mocked. It makes unit testing the services easier as well because then you'd need to simply mock the environment variable service and not test it over again in multiple places.

Reply:
I understand the motivation. 

But having a class that basically has the following method is a bit of an overkill imo.

```cs
private void GetEnvironmentVariablesForFeature(NuGetFeatureFlagConstants featureFlag, out bool isFeatureForcedEnabled, out bool isFeatureForcedDisabled)
        {
            isFeatureForcedEnabled = false;
            isFeatureForcedDisabled = false;
            if (!string.IsNullOrEmpty(featureFlag.EnvironmentVariable))
            {
                string envVarOverride = _environmentVariableReader.GetEnvironmentVariable(featureFlag.EnvironmentVariable);

                isFeatureForcedDisabled = envVarOverride == "0";
                isFeatureForcedEnabled = envVarOverride == "1";
            }
        }
```

> It makes unit testing the services easier as well because then you'd need to simply mock the environment override object and not test it over again in multiple places.

I'd argue these tests would need to be stay as is anyways, because you'd need to make sure that the correct environment variables are being read.